{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas\n",
    "from auxilary_functions import *\n",
    "import pickle\n",
    "import pandas as pd\n",
    "from os import listdir\n",
    "from os.path import isfile, join\n",
    "import re, sqlite3, gzip, os\n",
    "from timeit import default_timer as timer\n",
    "from sqlalchemy import create_engine"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_patent_num(jstr):\n",
    "    import re\n",
    "    jstr = str(jstr)\n",
    "    pats = re.findall('(: .+?(?=(\\s\\d*-\\w\\d*)))', jstr)\n",
    "    if not pats:\n",
    "        #print(pats)\n",
    "        return None\n",
    "    else:\n",
    "        pats = pats[0]\n",
    "        joiner = \"\".join\n",
    "        return \"\".join(list(filter(None, [joiner(words) for words in pats])))[2::].replace(\" \", \"\").replace(\"-\", \"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#genbank_db_filename = '/home/erikz/data/agnes_pkl_files/erik_ori_db_filtered.pkl'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/user/1024/ipykernel_1311856/1900663169.py:1: DtypeWarning: Columns (9,10,12,19,20,21,22) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  mabpat_db_uncleaned = pd.read_csv('/home/erikz/data/mabpat/dbs/MABPAT_09102023_COMPLETED_reserve.csv')\n"
     ]
    }
   ],
   "source": [
    "mabpat_db_uncleaned = pd.read_csv('/home/erikz/data/mabpat/dbs/MABPAT_09102023_COMPLETED_reserve.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Applicant</th>\n",
       "      <th>Applicant type</th>\n",
       "      <th>Applicant country</th>\n",
       "      <th>Year of application</th>\n",
       "      <th>Application number</th>\n",
       "      <th>Patent system</th>\n",
       "      <th>Sequence id</th>\n",
       "      <th>Species name</th>\n",
       "      <th>Domain</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>...</th>\n",
       "      <th>Sequence status</th>\n",
       "      <th>Is protein-coding sequence</th>\n",
       "      <th>Source of prediction</th>\n",
       "      <th>f_header</th>\n",
       "      <th>sseqid</th>\n",
       "      <th>stitle</th>\n",
       "      <th>pident</th>\n",
       "      <th>evalue</th>\n",
       "      <th>qcovs</th>\n",
       "      <th>appearance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BASF</td>\n",
       "      <td>MULTINATIONAL</td>\n",
       "      <td>GERMANY</td>\n",
       "      <td>2004</td>\n",
       "      <td>WO2004024931A2</td>\n",
       "      <td>WIPO</td>\n",
       "      <td>3418000</td>\n",
       "      <td>Vibrio cholerae</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>...</td>\n",
       "      <td>observed</td>\n",
       "      <td>True</td>\n",
       "      <td>TrEMBL</td>\n",
       "      <td>3418000_vibrio</td>\n",
       "      <td>tr|A0A0H3AH49|A0A0H3AH49_VIBC3</td>\n",
       "      <td>tr|A0A0H3AH49|A0A0H3AH49_VIBC3 Methylenetetrah...</td>\n",
       "      <td>99.7</td>\n",
       "      <td>2.670000e-223</td>\n",
       "      <td>99.7</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>YALE UNIVERSITY</td>\n",
       "      <td>UNIVERSITY</td>\n",
       "      <td>US</td>\n",
       "      <td>2010</td>\n",
       "      <td>JP2010213712A</td>\n",
       "      <td>JPO</td>\n",
       "      <td>3578959</td>\n",
       "      <td>Vibrio cholerae</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>...</td>\n",
       "      <td>observed</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>YALE UNIVERSITY</td>\n",
       "      <td>UNIVERSITY</td>\n",
       "      <td>US</td>\n",
       "      <td>2010</td>\n",
       "      <td>JP2010213712A</td>\n",
       "      <td>JPO</td>\n",
       "      <td>3578960</td>\n",
       "      <td>Vibrio cholerae</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>...</td>\n",
       "      <td>observed</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         Applicant Applicant type Applicant country  Year of application  \\\n",
       "0             BASF  MULTINATIONAL           GERMANY                 2004   \n",
       "1  YALE UNIVERSITY     UNIVERSITY                US                 2010   \n",
       "2  YALE UNIVERSITY     UNIVERSITY                US                 2010   \n",
       "\n",
       "  Application number Patent system  Sequence id     Species name    Domain  \\\n",
       "0     WO2004024931A2          WIPO      3418000  Vibrio cholerae  Bacteria   \n",
       "1      JP2010213712A           JPO      3578959  Vibrio cholerae  Bacteria   \n",
       "2      JP2010213712A           JPO      3578960  Vibrio cholerae  Bacteria   \n",
       "\n",
       "           Phylum  ... Sequence status  Is protein-coding sequence  \\\n",
       "0  Proteobacteria  ...        observed                        True   \n",
       "1  Proteobacteria  ...        observed                       False   \n",
       "2  Proteobacteria  ...        observed                       False   \n",
       "\n",
       "  Source of prediction        f_header                          sseqid  \\\n",
       "0               TrEMBL  3418000_vibrio  tr|A0A0H3AH49|A0A0H3AH49_VIBC3   \n",
       "1                  NaN             NaN                             NaN   \n",
       "2                  NaN             NaN                             NaN   \n",
       "\n",
       "                                              stitle  pident         evalue  \\\n",
       "0  tr|A0A0H3AH49|A0A0H3AH49_VIBC3 Methylenetetrah...    99.7  2.670000e-223   \n",
       "1                                                NaN     NaN            NaN   \n",
       "2                                                NaN     NaN            NaN   \n",
       "\n",
       "   qcovs appearance  \n",
       "0   99.7          1  \n",
       "1    NaN          1  \n",
       "2    NaN          1  \n",
       "\n",
       "[3 rows x 27 columns]"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mabpat_db_uncleaned.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Applicant', 'Applicant type', 'Applicant country',\n",
       "       'Year of application', 'Application number', 'Patent system',\n",
       "       'Sequence id', 'Species name', 'Domain', 'Phylum', 'Taxonomic source',\n",
       "       'Is Deep-sea', 'Deep-sea source', 'Sequence', 'GC content',\n",
       "       'Sequence length', 'Is marine sequence', 'Sequence status',\n",
       "       'Is protein-coding sequence', 'Source of prediction', 'f_header',\n",
       "       'sseqid', 'stitle', 'pident', 'evalue', 'qcovs', 'appearance'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mabpat_db_uncleaned.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "#mabpat_db_uncleaned = mabpat_db_uncleaned[mabpat_db_uncleaned.duplicated()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Applicant</th>\n",
       "      <th>Applicant type</th>\n",
       "      <th>Applicant country</th>\n",
       "      <th>Year of application</th>\n",
       "      <th>Application number</th>\n",
       "      <th>Patent system</th>\n",
       "      <th>Sequence id</th>\n",
       "      <th>Species name</th>\n",
       "      <th>Domain</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>...</th>\n",
       "      <th>Sequence status</th>\n",
       "      <th>Is protein-coding sequence</th>\n",
       "      <th>Source of prediction</th>\n",
       "      <th>f_header</th>\n",
       "      <th>sseqid</th>\n",
       "      <th>stitle</th>\n",
       "      <th>pident</th>\n",
       "      <th>evalue</th>\n",
       "      <th>qcovs</th>\n",
       "      <th>appearance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>BASF</td>\n",
       "      <td>MULTINATIONAL</td>\n",
       "      <td>GERMANY</td>\n",
       "      <td>2004</td>\n",
       "      <td>WO2004024931A2</td>\n",
       "      <td>WIPO</td>\n",
       "      <td>3418000</td>\n",
       "      <td>Vibrio cholerae</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>...</td>\n",
       "      <td>observed</td>\n",
       "      <td>True</td>\n",
       "      <td>TrEMBL</td>\n",
       "      <td>3418000_vibrio</td>\n",
       "      <td>tr|A0A0H3AH49|A0A0H3AH49_VIBC3</td>\n",
       "      <td>tr|A0A0H3AH49|A0A0H3AH49_VIBC3 Methylenetetrah...</td>\n",
       "      <td>99.7</td>\n",
       "      <td>2.670000e-223</td>\n",
       "      <td>99.7</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>YALE UNIVERSITY</td>\n",
       "      <td>UNIVERSITY</td>\n",
       "      <td>US</td>\n",
       "      <td>2010</td>\n",
       "      <td>JP2010213712A</td>\n",
       "      <td>JPO</td>\n",
       "      <td>3578959</td>\n",
       "      <td>Vibrio cholerae</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>...</td>\n",
       "      <td>observed</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>YALE UNIVERSITY</td>\n",
       "      <td>UNIVERSITY</td>\n",
       "      <td>US</td>\n",
       "      <td>2010</td>\n",
       "      <td>JP2010213712A</td>\n",
       "      <td>JPO</td>\n",
       "      <td>3578960</td>\n",
       "      <td>Vibrio cholerae</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>...</td>\n",
       "      <td>observed</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>YALE UNIVERSITY</td>\n",
       "      <td>UNIVERSITY</td>\n",
       "      <td>US</td>\n",
       "      <td>2010</td>\n",
       "      <td>JP2010213712A</td>\n",
       "      <td>JPO</td>\n",
       "      <td>3578961</td>\n",
       "      <td>Vibrio cholerae</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>...</td>\n",
       "      <td>observed</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>YALE UNIVERSITY</td>\n",
       "      <td>UNIVERSITY</td>\n",
       "      <td>US</td>\n",
       "      <td>2010</td>\n",
       "      <td>JP2010213712A</td>\n",
       "      <td>JPO</td>\n",
       "      <td>3579196</td>\n",
       "      <td>Vibrio cholerae</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>...</td>\n",
       "      <td>observed</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>692348</th>\n",
       "      <td>GENIZON BIOSCIENCES INC [CA]</td>\n",
       "      <td>NATIONAL</td>\n",
       "      <td>CANADA</td>\n",
       "      <td>2009</td>\n",
       "      <td>EP2126131A2</td>\n",
       "      <td>EPO</td>\n",
       "      <td>54136045</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>predicted</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>692349</th>\n",
       "      <td>BAYER</td>\n",
       "      <td>MULTINATIONAL</td>\n",
       "      <td>GERMANY</td>\n",
       "      <td>2010</td>\n",
       "      <td>WO2010121818A1</td>\n",
       "      <td>WIPO</td>\n",
       "      <td>54270389</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>predicted</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>692350</th>\n",
       "      <td>BAYER</td>\n",
       "      <td>MULTINATIONAL</td>\n",
       "      <td>GERMANY</td>\n",
       "      <td>2010</td>\n",
       "      <td>WO2010121818A1</td>\n",
       "      <td>WIPO</td>\n",
       "      <td>54270390</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "      <td>Nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>predicted</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>692351</th>\n",
       "      <td>BAYER</td>\n",
       "      <td>MULTINATIONAL</td>\n",
       "      <td>GERMANY</td>\n",
       "      <td>2010</td>\n",
       "      <td>WO2010121818A1</td>\n",
       "      <td>WIPO</td>\n",
       "      <td>54270391</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>predicted</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>692352</th>\n",
       "      <td>BAYER</td>\n",
       "      <td>MULTINATIONAL</td>\n",
       "      <td>GERMANY</td>\n",
       "      <td>2010</td>\n",
       "      <td>WO2010121818A1</td>\n",
       "      <td>WIPO</td>\n",
       "      <td>54270392</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Nan</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>predicted</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>692353 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                           Applicant Applicant type Applicant country  \\\n",
       "0                               BASF  MULTINATIONAL           GERMANY   \n",
       "1                    YALE UNIVERSITY     UNIVERSITY                US   \n",
       "2                    YALE UNIVERSITY     UNIVERSITY                US   \n",
       "3                    YALE UNIVERSITY     UNIVERSITY                US   \n",
       "4                    YALE UNIVERSITY     UNIVERSITY                US   \n",
       "...                              ...            ...               ...   \n",
       "692348  GENIZON BIOSCIENCES INC [CA]       NATIONAL            CANADA   \n",
       "692349                         BAYER  MULTINATIONAL           GERMANY   \n",
       "692350                         BAYER  MULTINATIONAL           GERMANY   \n",
       "692351                         BAYER  MULTINATIONAL           GERMANY   \n",
       "692352                         BAYER  MULTINATIONAL           GERMANY   \n",
       "\n",
       "        Year of application Application number Patent system  Sequence id  \\\n",
       "0                      2004     WO2004024931A2          WIPO      3418000   \n",
       "1                      2010      JP2010213712A           JPO      3578959   \n",
       "2                      2010      JP2010213712A           JPO      3578960   \n",
       "3                      2010      JP2010213712A           JPO      3578961   \n",
       "4                      2010      JP2010213712A           JPO      3579196   \n",
       "...                     ...                ...           ...          ...   \n",
       "692348                 2009        EP2126131A2           EPO     54136045   \n",
       "692349                 2010     WO2010121818A1          WIPO     54270389   \n",
       "692350                 2010     WO2010121818A1          WIPO     54270390   \n",
       "692351                 2010     WO2010121818A1          WIPO     54270391   \n",
       "692352                 2010     WO2010121818A1          WIPO     54270392   \n",
       "\n",
       "                Species name    Domain          Phylum  ... Sequence status  \\\n",
       "0            Vibrio cholerae  Bacteria  Proteobacteria  ...        observed   \n",
       "1            Vibrio cholerae  Bacteria  Proteobacteria  ...        observed   \n",
       "2            Vibrio cholerae  Bacteria  Proteobacteria  ...        observed   \n",
       "3            Vibrio cholerae  Bacteria  Proteobacteria  ...        observed   \n",
       "4            Vibrio cholerae  Bacteria  Proteobacteria  ...        observed   \n",
       "...                      ...       ...             ...  ...             ...   \n",
       "692348          Homo sapiens       Nan             NaN  ...       predicted   \n",
       "692349  Arabidopsis thaliana       Nan             NaN  ...       predicted   \n",
       "692350  Arabidopsis thaliana       Nan             NaN  ...       predicted   \n",
       "692351          Homo sapiens       Nan             NaN  ...       predicted   \n",
       "692352          Homo sapiens       Nan             NaN  ...       predicted   \n",
       "\n",
       "        Is protein-coding sequence Source of prediction        f_header  \\\n",
       "0                             True               TrEMBL  3418000_vibrio   \n",
       "1                            False                  NaN             NaN   \n",
       "2                            False                  NaN             NaN   \n",
       "3                            False                  NaN             NaN   \n",
       "4                            False                  NaN             NaN   \n",
       "...                            ...                  ...             ...   \n",
       "692348                       False                  NaN             NaN   \n",
       "692349                       False                  NaN             NaN   \n",
       "692350                        True                  NaN             NaN   \n",
       "692351                        True                  NaN             NaN   \n",
       "692352                        True                  NaN             NaN   \n",
       "\n",
       "                                sseqid  \\\n",
       "0       tr|A0A0H3AH49|A0A0H3AH49_VIBC3   \n",
       "1                                  NaN   \n",
       "2                                  NaN   \n",
       "3                                  NaN   \n",
       "4                                  NaN   \n",
       "...                                ...   \n",
       "692348                             NaN   \n",
       "692349                             NaN   \n",
       "692350                             NaN   \n",
       "692351                             NaN   \n",
       "692352                             NaN   \n",
       "\n",
       "                                                   stitle  pident  \\\n",
       "0       tr|A0A0H3AH49|A0A0H3AH49_VIBC3 Methylenetetrah...    99.7   \n",
       "1                                                     NaN     NaN   \n",
       "2                                                     NaN     NaN   \n",
       "3                                                     NaN     NaN   \n",
       "4                                                     NaN     NaN   \n",
       "...                                                   ...     ...   \n",
       "692348                                                NaN     NaN   \n",
       "692349                                                NaN     NaN   \n",
       "692350                                                NaN     NaN   \n",
       "692351                                                NaN     NaN   \n",
       "692352                                                NaN     NaN   \n",
       "\n",
       "               evalue  qcovs appearance  \n",
       "0       2.670000e-223   99.7          1  \n",
       "1                 NaN    NaN          1  \n",
       "2                 NaN    NaN          1  \n",
       "3                 NaN    NaN          1  \n",
       "4                 NaN    NaN          1  \n",
       "...               ...    ...        ...  \n",
       "692348            NaN    NaN          1  \n",
       "692349            NaN    NaN          1  \n",
       "692350            NaN    NaN          1  \n",
       "692351            NaN    NaN          1  \n",
       "692352            NaN    NaN          1  \n",
       "\n",
       "[692353 rows x 27 columns]"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mabpat_db_uncleaned"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parse accession number"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "DB = '/home/erikz/data/patents_db/genepat_Db_V4_with_accession.sqlite'\n",
    "FILEDIR = '/home/erikz/data/patents_db/genbank_data_V4/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_genefiles = sorted([f for f in os.listdir(FILEDIR) if f.endswith('.seq.gz')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_genefiles = all_genefiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_journal(seq):\n",
    "    \n",
    "    try:\n",
    "        ix_j = seq.index('JOURNAL')\n",
    "        ix_f = seq.index('FEATURES')\n",
    "        \n",
    "        journal = seq[ix_j:ix_f].strip()\n",
    "        \n",
    "    except:\n",
    "        return None # Returns None if substring doesnt exist\n",
    "    \n",
    "    return journal"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_origin(seq):\n",
    "    \n",
    "    try:\n",
    "        ix_j = seq.index('ORIGIN ')\n",
    "        \n",
    "        origin = seq[ix_j:].split('\\n')\n",
    "        origin = ''.join([line[10:].replace(' ','') for line in origin[1:]])\n",
    "        \n",
    "    except:\n",
    "        return None # Returns None if substring doesnt exist\n",
    "    \n",
    "    return origin"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Create table for Db\n",
    "conn = sqlite3.connect(DB)\n",
    "cur = conn.cursor()\n",
    "\n",
    "## SQL used to create db:\n",
    "cur.execute('''CREATE TABLE \"patents_accession\" (\"id\" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL UNIQUE, \"accession\" TEXT, \"locus\" TEXT, \"organism\" TEXT, \"definition\" TEXT, \"journal\" TEXT, \"filename\" TEXT, \"origin\" TEXT)''')    \n",
    "\n",
    "conn.commit()\n",
    "conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing file...: gbpat1.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat10.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat100.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat101.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat102.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat103.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat104.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat105.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat106.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat107.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat108.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat109.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat11.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat110.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat111.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat112.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat113.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat114.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat115.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat116.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat117.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat118.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat119.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat12.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat120.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat121.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat122.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat123.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat124.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat125.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat126.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat127.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat128.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat129.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat13.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat130.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat131.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat132.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat133.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat134.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat135.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat136.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat137.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat138.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat139.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat14.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat140.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat141.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat142.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat143.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat144.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat145.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat146.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat147.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat148.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat149.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat15.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat150.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat151.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat152.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat153.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat154.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat155.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat156.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat157.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat158.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat159.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat16.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat160.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat161.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat162.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat163.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat164.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat165.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat166.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat167.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat168.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat169.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat17.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat170.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat171.seq.gz\n",
      "Last post removed\n",
      "Last post removed\n",
      "Processing file...: gbpat177.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat178.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat179.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat18.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat180.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat181.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat182.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat183.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat184.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat185.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat186.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat187.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat188.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat189.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat19.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat190.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat191.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat192.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat193.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat194.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat195.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat196.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat197.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat198.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat199.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat2.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat20.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat200.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat201.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat202.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat203.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat204.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat205.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat206.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat207.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat208.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat209.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat21.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat210.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat211.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat212.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat213.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat214.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat215.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat216.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat217.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat218.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat219.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat22.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat220.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat221.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat222.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat223.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat224.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat225.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat226.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat227.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat228.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat229.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat23.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat230.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat231.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat232.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat233.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat234.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat235.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat236.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat237.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat238.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat239.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat24.seq.gz\n",
      "Last post removed\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing file...: gbpat240.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat241.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat242.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat243.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat244.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat245.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat246.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat247.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat248.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat249.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat25.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat250.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat251.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat26.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat27.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat28.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat29.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat3.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat30.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat31.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat32.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat33.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat34.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat35.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat36.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat37.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat38.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat39.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat4.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat40.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat41.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat42.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat43.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat44.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat45.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat46.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat47.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat48.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat49.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat5.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat50.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat51.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat52.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat53.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat54.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat55.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat56.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat57.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat58.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat59.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat6.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat60.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat61.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat62.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat63.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat64.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat65.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat66.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat67.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat68.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat69.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat7.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat70.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat71.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat72.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat73.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat74.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat75.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat76.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat77.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat78.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat79.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat8.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat80.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat81.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat82.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat83.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat84.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat85.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat86.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat87.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat88.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat89.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat9.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat90.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat91.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat92.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat93.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat94.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat95.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat96.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat97.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat98.seq.gz\n",
      "Last post removed\n",
      "Processing file...: gbpat99.seq.gz\n",
      "Last post removed\n",
      "The total runtime is 5591.966869645985  seconds.\n"
     ]
    }
   ],
   "source": [
    "## For each file - insert it´s contents to db \n",
    "start1 = timer()\n",
    "for genefile in all_genefiles: \n",
    "\n",
    "    print('Processing file...: %s' % genefile)\n",
    "    \n",
    "    # Read zipped file while decoding from bytes to string\n",
    "    f = gzip.open(FILEDIR + genefile,'rb')\n",
    "    ff = f.read().decode()\n",
    "\n",
    "    # Split up file into separate patent chunks\n",
    "    sequences = ff.split('\\n//\\n') # Yes, it gets the first one correct too! \n",
    "    # ff.count('//') # nr of separators... should be -1 from len of sequences\n",
    "\n",
    "    # Grab out the data we want \n",
    "    ll = []\n",
    "\n",
    "    for seq in sequences:\n",
    "        lines = seq.splitlines()[:]  # ORGANISM, JOURNAL, DEFINITION\n",
    "        lines = [line.strip() for line in lines]\n",
    "\n",
    "        ll.append( [\n",
    "            ''.join([line for line in lines if line.startswith('ACCESSION')]),\n",
    "            ''.join([line for line in lines if line.startswith('LOCUS')]), \n",
    "            ''.join([line for line in lines if line.startswith('ORGANISM')]), \n",
    "            ''.join([line for line in lines if line.startswith('DEFINITION')]),\n",
    "            #''.join([line for line in lines if line.startswith('JOURNAL')]),  \n",
    "            get_journal(seq), # Get out JOURNAL section using helper method above \n",
    "            genefile,\n",
    "            get_origin(seq)\n",
    "                ]\n",
    "        )\n",
    "\n",
    "    if ll[-1][0] == '': # if the last one is empty (probably always...), then remove it\n",
    "        ll = ll[:-1]\n",
    "        print('Last post removed')\n",
    "        \n",
    "    # Add data to db\n",
    "    # tbl: patent \n",
    "    conn = sqlite3.connect(DB)\n",
    "    cur = conn.cursor()\n",
    "    \n",
    "    cur.executemany('insert into patents_accession(accession, locus, organism, definition, journal, filename, origin) values (?,?,?,?,?,?,?)', ll)\n",
    "    conn.commit()\n",
    "    conn.close()\n",
    "    \n",
    "end1 = timer()\n",
    "print('The total runtime is',(end1-start1),' seconds.')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create DB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The total runtime is 466.54978655907325  seconds.\n"
     ]
    }
   ],
   "source": [
    "start1 = timer()\n",
    "engine = create_engine('sqlite:///%s' % DB)\n",
    "\n",
    "df_all_w_accession = pd.read_sql_query(\"SELECT id, accession, organism, journal, filename, origin FROM patents_accession;\", engine)\n",
    "end1 = timer()\n",
    "print('The total runtime is',(end1-start1),' seconds.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>accession</th>\n",
       "      <th>organism</th>\n",
       "      <th>journal</th>\n",
       "      <th>filename</th>\n",
       "      <th>origin</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ACCESSION   A00001</td>\n",
       "      <td>ORGANISM  Cauliflower mosaic virus</td>\n",
       "      <td>JOURNAL   Patent: EP 0242016-A1 1 21-OCT-1987;...</td>\n",
       "      <td>gbpat1.seq.gz</td>\n",
       "      <td>gttttgtttgatggagaattgcgcagaggggttatatctgcgtgag...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>ACCESSION   A00002</td>\n",
       "      <td>ORGANISM  Bos taurus</td>\n",
       "      <td>JOURNAL   Patent: EP 0238993-A2 1 30-SEP-1987;...</td>\n",
       "      <td>gbpat1.seq.gz</td>\n",
       "      <td>aattcatgcgtccggacttctgcctcgagccgccgtacactgggcc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>ACCESSION   A00003</td>\n",
       "      <td>ORGANISM  Bos taurus</td>\n",
       "      <td>JOURNAL   Patent: EP 0238993-A2 2 30-SEP-1987;...</td>\n",
       "      <td>gbpat1.seq.gz</td>\n",
       "      <td>caagctttactaaggaccaccgcaagtacgttcgcagtcttccgcg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>ACCESSION   A00004</td>\n",
       "      <td>ORGANISM  synthetic construct</td>\n",
       "      <td>JOURNAL   Patent: EP 0238993-A2 3 30-SEP-1987;...</td>\n",
       "      <td>gbpat1.seq.gz</td>\n",
       "      <td>aattcatgcgtccggacttctgcctcgagc</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>ACCESSION   A00005</td>\n",
       "      <td>ORGANISM  synthetic construct</td>\n",
       "      <td>JOURNAL   Patent: EP 0238993-A2 4 30-SEP-1987;...</td>\n",
       "      <td>gbpat1.seq.gz</td>\n",
       "      <td>cagaagtccggacgcatg</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55331419</th>\n",
       "      <td>55331420</td>\n",
       "      <td>ACCESSION   HJ966292</td>\n",
       "      <td>ORGANISM  Unknown.</td>\n",
       "      <td>JOURNAL   Patent: US 8722325-B2 29 13-MAY-2014...</td>\n",
       "      <td>gbpat99.seq.gz</td>\n",
       "      <td>taaagtaccacacatcttgtttattggatattgcctttctt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55331420</th>\n",
       "      <td>55331421</td>\n",
       "      <td>ACCESSION   HJ966293</td>\n",
       "      <td>ORGANISM  Unknown.</td>\n",
       "      <td>JOURNAL   Patent: US 8722325-B2 30 13-MAY-2014...</td>\n",
       "      <td>gbpat99.seq.gz</td>\n",
       "      <td>atatgatcaggacatgtggttccattgatagcccaacttctt</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55331421</th>\n",
       "      <td>55331422</td>\n",
       "      <td>ACCESSION   HJ966294</td>\n",
       "      <td>ORGANISM  Unknown.</td>\n",
       "      <td>JOURNAL   Patent: US 8722326-B2 1 13-MAY-2014;...</td>\n",
       "      <td>gbpat99.seq.gz</td>\n",
       "      <td>nnnnnnnngcatancacgangtcatnatcgtncaaacgtcagtcca...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55331422</th>\n",
       "      <td>55331423</td>\n",
       "      <td>ACCESSION   HJ966295</td>\n",
       "      <td>ORGANISM  Unknown.</td>\n",
       "      <td>JOURNAL   Patent: US 8722326-B2 2 13-MAY-2014;...</td>\n",
       "      <td>gbpat99.seq.gz</td>\n",
       "      <td>tatcatctggatgttaggaagacaaaaggaagctgaggacattaac...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55331423</th>\n",
       "      <td>55331424</td>\n",
       "      <td>ACCESSION   HJ966296</td>\n",
       "      <td>ORGANISM  Unknown.</td>\n",
       "      <td>JOURNAL   Patent: US 8722326-B2 3 13-MAY-2014;...</td>\n",
       "      <td>gbpat99.seq.gz</td>\n",
       "      <td>accttcagaccagat</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>55331424 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                id             accession                            organism  \\\n",
       "0                1    ACCESSION   A00001  ORGANISM  Cauliflower mosaic virus   \n",
       "1                2    ACCESSION   A00002                ORGANISM  Bos taurus   \n",
       "2                3    ACCESSION   A00003                ORGANISM  Bos taurus   \n",
       "3                4    ACCESSION   A00004       ORGANISM  synthetic construct   \n",
       "4                5    ACCESSION   A00005       ORGANISM  synthetic construct   \n",
       "...            ...                   ...                                 ...   \n",
       "55331419  55331420  ACCESSION   HJ966292                  ORGANISM  Unknown.   \n",
       "55331420  55331421  ACCESSION   HJ966293                  ORGANISM  Unknown.   \n",
       "55331421  55331422  ACCESSION   HJ966294                  ORGANISM  Unknown.   \n",
       "55331422  55331423  ACCESSION   HJ966295                  ORGANISM  Unknown.   \n",
       "55331423  55331424  ACCESSION   HJ966296                  ORGANISM  Unknown.   \n",
       "\n",
       "                                                    journal        filename  \\\n",
       "0         JOURNAL   Patent: EP 0242016-A1 1 21-OCT-1987;...   gbpat1.seq.gz   \n",
       "1         JOURNAL   Patent: EP 0238993-A2 1 30-SEP-1987;...   gbpat1.seq.gz   \n",
       "2         JOURNAL   Patent: EP 0238993-A2 2 30-SEP-1987;...   gbpat1.seq.gz   \n",
       "3         JOURNAL   Patent: EP 0238993-A2 3 30-SEP-1987;...   gbpat1.seq.gz   \n",
       "4         JOURNAL   Patent: EP 0238993-A2 4 30-SEP-1987;...   gbpat1.seq.gz   \n",
       "...                                                     ...             ...   \n",
       "55331419  JOURNAL   Patent: US 8722325-B2 29 13-MAY-2014...  gbpat99.seq.gz   \n",
       "55331420  JOURNAL   Patent: US 8722325-B2 30 13-MAY-2014...  gbpat99.seq.gz   \n",
       "55331421  JOURNAL   Patent: US 8722326-B2 1 13-MAY-2014;...  gbpat99.seq.gz   \n",
       "55331422  JOURNAL   Patent: US 8722326-B2 2 13-MAY-2014;...  gbpat99.seq.gz   \n",
       "55331423  JOURNAL   Patent: US 8722326-B2 3 13-MAY-2014;...  gbpat99.seq.gz   \n",
       "\n",
       "                                                     origin  \n",
       "0         gttttgtttgatggagaattgcgcagaggggttatatctgcgtgag...  \n",
       "1         aattcatgcgtccggacttctgcctcgagccgccgtacactgggcc...  \n",
       "2         caagctttactaaggaccaccgcaagtacgttcgcagtcttccgcg...  \n",
       "3                            aattcatgcgtccggacttctgcctcgagc  \n",
       "4                                        cagaagtccggacgcatg  \n",
       "...                                                     ...  \n",
       "55331419          taaagtaccacacatcttgtttattggatattgcctttctt  \n",
       "55331420         atatgatcaggacatgtggttccattgatagcccaacttctt  \n",
       "55331421  nnnnnnnngcatancacgangtcatnatcgtncaaacgtcagtcca...  \n",
       "55331422  tatcatctggatgttaggaagacaaaaggaagctgaggacattaac...  \n",
       "55331423                                    accttcagaccagat  \n",
       "\n",
       "[55331424 rows x 6 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all_w_accession"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all_w_accession['accession_num'] = df_all_w_accession.accession.apply(lambda x: x[9:].strip())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all_w_accession['species_name'] = df_all_w_accession.organism.apply(lambda x: x[9:].strip())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all_w_accession['Sequence id'] = df_all_w_accession['id']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all_w_accession['journal'].replace('', np.nan, inplace=True) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all_w_accession['patent_num'] = df_all_w_accession.journal.apply(lambda x: get_patent_num(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['id', 'accession', 'organism', 'journal', 'filename', 'origin',\n",
       "       'accession_num', 'Species name', 'species_name', 'Sequence id',\n",
       "       'patent_num'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all_w_accession.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession_num</th>\n",
       "      <th>patent_num</th>\n",
       "      <th>species_name</th>\n",
       "      <th>origin</th>\n",
       "      <th>Sequence id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A00001</td>\n",
       "      <td>EP0242016A1</td>\n",
       "      <td>Cauliflower mosaic virus</td>\n",
       "      <td>gttttgtttgatggagaattgcgcagaggggttatatctgcgtgag...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>A00002</td>\n",
       "      <td>EP0238993A2</td>\n",
       "      <td>Bos taurus</td>\n",
       "      <td>aattcatgcgtccggacttctgcctcgagccgccgtacactgggcc...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A00003</td>\n",
       "      <td>EP0238993A2</td>\n",
       "      <td>Bos taurus</td>\n",
       "      <td>caagctttactaaggaccaccgcaagtacgttcgcagtcttccgcg...</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  accession_num   patent_num              species_name  \\\n",
       "0        A00001  EP0242016A1  Cauliflower mosaic virus   \n",
       "1        A00002  EP0238993A2                Bos taurus   \n",
       "2        A00003  EP0238993A2                Bos taurus   \n",
       "\n",
       "                                              origin  Sequence id  \n",
       "0  gttttgtttgatggagaattgcgcagaggggttatatctgcgtgag...            1  \n",
       "1  aattcatgcgtccggacttctgcctcgagccgccgtacactgggcc...            2  \n",
       "2  caagctttactaaggaccaccgcaagtacgttcgcagtcttccgcg...            3  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_all_w_accession_save = df_all_w_accession[['accession_num', 'patent_num', 'species_name', 'origin', 'Sequence id']]\n",
    "df_all_w_accession_save.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/home/erikz/data/patents_db/genepat_Db_V4_with_accession.pkl', 'wb') as f:\n",
    "    pickle.dump(df_all_w_accession_save, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "genbak_unique_patent_nums = df_all_w_accession_save.patent_num.drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "genbak_unique_patent_nums.to_csv('/home/erikz/data/patents_db/genepat_Db_V4_unique_patent_nums.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Add accession numbers to MABPAT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_all_w_accession_save = pd.read_pickle('/home/erikz/data/patents_db/genepat_Db_V4_with_accession.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [],
   "source": [
    "mabpat_db = pd.merge(mabpat_db_uncleaned,df_all_w_accession_save,how='inner',on='Sequence id').drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [],
   "source": [
    "#mabpat_db_uncleaned"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [],
   "source": [
    "mabpat_db_save = mabpat_db[['Applicant', 'Applicant type', 'Applicant country',\n",
    "       'Year of application', 'Application number', 'Patent system',\n",
    "       'Sequence id', 'Species name', 'Domain', 'Phylum', 'Taxonomic source',\n",
    "       'Is Deep-sea', 'Deep-sea source', 'Sequence', 'GC content',\n",
    "       'Sequence length', 'Is marine sequence', 'Sequence status',\n",
    "       'Is protein-coding sequence', 'Source of prediction', 'f_header',\n",
    "       'sseqid', 'stitle', 'pident', 'evalue', 'qcovs', 'appearance',\n",
    "       'accession_num']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Applicant', 'Applicant type', 'Applicant country',\n",
       "       'Year of application', 'Application number', 'Patent system',\n",
       "       'Sequence id', 'Species name', 'Domain', 'Phylum', 'Taxonomic source',\n",
       "       'Is Deep-sea', 'Deep-sea source', 'Sequence', 'GC content',\n",
       "       'Sequence length', 'Is marine sequence', 'Sequence status',\n",
       "       'Is protein-coding sequence', 'Source of prediction', 'f_header',\n",
       "       'sseqid', 'stitle', 'pident', 'evalue', 'qcovs', 'appearance',\n",
       "       'accession_num'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 142,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mabpat_db_save.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [],
   "source": [
    "mabpat_db_save.to_csv('/home/erikz/data/mabpat/dbs/MABPAT_09102023_COMPLETED_with_accession.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Create final sql"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/user/1024/ipykernel_3420031/1908186986.py:1: DtypeWarning: Columns (9,10,12,19,20,21,22) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  mabpat_db_cleaned = pd.read_csv('/home/erikz/data/mabpat/dbs/MABPAT_09102023_COMPLETED_with_accession.csv')\n"
     ]
    }
   ],
   "source": [
    "mabpat_db_cleaned = pd.read_csv('/home/erikz/data/mabpat/dbs/MABPAT_09102023_COMPLETED_with_accession.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Save fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "marine_df_for_blast = mabpat_db_cleaned.drop_duplicates(subset='Sequence', keep='first')\n",
    "marine_df_for_blast = marine_df_for_blast.sort_values(by='Species name', ascending=True)\n",
    "#test_df = marine_df.sample(200)\n",
    "marine_df_for_blast['fasta_entry'] = '>'+marine_df_for_blast['accession_num'].apply(str)+'_'+marine_df_for_blast['Species name']+'\\n'+marine_df_for_blast['Sequence']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"/home/erikz/data/mabpat/dbs/mabpat_nucleotide_sequences.fa\", \"w\") as f_out:\n",
    "    f_out.write(\"\\n\".join(marine_df_for_blast[\"fasta_entry\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#mabpat_db_cleaned = mabpat_db_save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Applicant', 'Applicant type', 'Applicant country',\n",
       "       'Year of application', 'Application number', 'Patent system',\n",
       "       'Sequence id', 'Species name', 'Domain', 'Phylum', 'Taxonomic source',\n",
       "       'Is Deep-sea', 'Deep-sea source', 'Sequence', 'GC content',\n",
       "       'Sequence length', 'Is marine sequence', 'Sequence status',\n",
       "       'Is protein-coding sequence', 'Source of prediction', 'f_header',\n",
       "       'sseqid', 'stitle', 'pident', 'evalue', 'qcovs', 'appearance',\n",
       "       'accession_num'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mabpat_db_cleaned.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "del mabpat_db_cleaned['Sequence']\n",
    "del mabpat_db_cleaned['appearance']\n",
    "#del mabpat_db_cleaned['Unnamed: 0']\n",
    "mabpat_db_cleaned['Is annotated'] = mabpat_db_cleaned['f_header'].notna().astype(int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Applicant', 'Applicant type', 'Applicant country',\n",
       "       'Year of application', 'Application number', 'Patent system',\n",
       "       'Sequence id', 'Species name', 'Domain', 'Phylum', 'Taxonomic source',\n",
       "       'Is Deep-sea', 'Deep-sea source', 'GC content', 'Sequence length',\n",
       "       'Is marine sequence', 'Sequence status', 'Is protein-coding sequence',\n",
       "       'Source of prediction', 'f_header', 'sseqid', 'stitle', 'pident',\n",
       "       'evalue', 'qcovs', 'accession_num', 'Is annotated'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mabpat_db_cleaned.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Drop all non-marine sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "mabpat_db_cleaned_marine = mabpat_db_cleaned[mabpat_db_cleaned['Is marine sequence']==1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "mabpat_db_cleaned_non_marine = mabpat_db_cleaned[mabpat_db_cleaned['Is marine sequence']==0]\n",
    "mabpat_db_cleaned_non_marine = mabpat_db_cleaned_non_marine[['accession_num', 'Species name']].drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#applicants table\n",
    "mabpat_db_applicants = mabpat_db_cleaned_marine[['Applicant', 'Applicant type', 'Applicant country']]\n",
    "\n",
    "#applictions table\n",
    "mabpat_db_applications = mabpat_db_cleaned_marine[['Application number', 'Applicant',\n",
    "                                          'Year of application', 'Patent system']]\n",
    "#species table\n",
    "mabpat_db_species = mabpat_db_cleaned_marine[['Species name', 'Domain', 'Phylum', 'Taxonomic source',\n",
    "                                       'Is Deep-sea', 'Deep-sea source', 'Is marine sequence']]\n",
    "#patents-sequences table\n",
    "mabpat_db_app_sequences = mabpat_db_cleaned[['Application number', 'accession_num']]\n",
    "\n",
    "#sequences table\n",
    "mabpat_db_sequences = mabpat_db_cleaned_marine[['accession_num', 'Species name', 'GC content', 'Sequence length',\n",
    "       'Is marine sequence', 'Sequence status', 'Is protein-coding sequence', 'Is annotated']]\n",
    "\n",
    "#functions table\n",
    "mabpat_db_marine_proteins_annotated_functions = mabpat_db_cleaned_marine[['accession_num', 'f_header', 'sseqid', 'stitle', 'pident',\n",
    "       'evalue', 'qcovs']]\n",
    "\n",
    "#non-marine sequences table\n",
    "#mabpat_db_cleaned_non_marine"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### applicants table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "mabpat_db_applicants = mabpat_db_applicants.sort_values(by='Applicant', ascending=True).drop_duplicates(subset='Applicant', keep='first')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Applicant</th>\n",
       "      <th>Applicant type</th>\n",
       "      <th>Applicant country</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>592538</th>\n",
       "      <td>1GLOBE BIOMEDICAL CO., LTD.|1GLOBE HEALTH INST...</td>\n",
       "      <td>NATIONAL</td>\n",
       "      <td>HONG KONG SAR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>535661</th>\n",
       "      <td>3-DIMENSIONAL PHARMACEUTICALS, INC.; EXTON, PA</td>\n",
       "      <td>NATIONAL</td>\n",
       "      <td>US</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11303</th>\n",
       "      <td>A GREY CRAIG,DAVID GRIFFEN,BALDOMERO M OLIVERA...</td>\n",
       "      <td>NONE</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534289</th>\n",
       "      <td>AARHUS UNIVERSITET (DK)</td>\n",
       "      <td>UNIVERSITY</td>\n",
       "      <td>DENMARK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16005</th>\n",
       "      <td>AARON KAPLAN,JUDY LIEMAN-HURWITZ,SHIMON RACHMI...</td>\n",
       "      <td>NONE</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37514</th>\n",
       "      <td>ZACHARY Q BECK,JEFFREY W MUNOS,DEREK H WELLS,J...</td>\n",
       "      <td>NONE</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30879</th>\n",
       "      <td>ZAKRYTOE AKTSIONERNOE OBSCHESTVO 'EVROGEN' (RU)</td>\n",
       "      <td>NONE</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534677</th>\n",
       "      <td>ZALTSMAN, ADI</td>\n",
       "      <td>NONE</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>579298</th>\n",
       "      <td>ZENSUN (SHANGHAI) SCI &amp; TECH CO., LTD</td>\n",
       "      <td>NATIONAL</td>\n",
       "      <td>CHINA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10670</th>\n",
       "      <td>ZUCHEM INC [US]</td>\n",
       "      <td>NATIONAL</td>\n",
       "      <td>US</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1326 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                Applicant Applicant type  \\\n",
       "592538  1GLOBE BIOMEDICAL CO., LTD.|1GLOBE HEALTH INST...       NATIONAL   \n",
       "535661     3-DIMENSIONAL PHARMACEUTICALS, INC.; EXTON, PA       NATIONAL   \n",
       "11303   A GREY CRAIG,DAVID GRIFFEN,BALDOMERO M OLIVERA...           NONE   \n",
       "534289                            AARHUS UNIVERSITET (DK)     UNIVERSITY   \n",
       "16005   AARON KAPLAN,JUDY LIEMAN-HURWITZ,SHIMON RACHMI...           NONE   \n",
       "...                                                   ...            ...   \n",
       "37514   ZACHARY Q BECK,JEFFREY W MUNOS,DEREK H WELLS,J...           NONE   \n",
       "30879     ZAKRYTOE AKTSIONERNOE OBSCHESTVO 'EVROGEN' (RU)           NONE   \n",
       "534677                                      ZALTSMAN, ADI           NONE   \n",
       "579298              ZENSUN (SHANGHAI) SCI & TECH CO., LTD       NATIONAL   \n",
       "10670                                     ZUCHEM INC [US]       NATIONAL   \n",
       "\n",
       "       Applicant country  \n",
       "592538     HONG KONG SAR  \n",
       "535661                US  \n",
       "11303                NaN  \n",
       "534289           DENMARK  \n",
       "16005                NaN  \n",
       "...                  ...  \n",
       "37514                NaN  \n",
       "30879                NaN  \n",
       "534677               NaN  \n",
       "579298             CHINA  \n",
       "10670                 US  \n",
       "\n",
       "[1326 rows x 3 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mabpat_db_applicants"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### applications table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Application number</th>\n",
       "      <th>Applicant</th>\n",
       "      <th>Year of application</th>\n",
       "      <th>Patent system</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>534802</th>\n",
       "      <td>DE19859107A1</td>\n",
       "      <td>ROCHE DIAGNOSTICS</td>\n",
       "      <td>2000</td>\n",
       "      <td>EPO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>579459</th>\n",
       "      <td>EP0130166A1</td>\n",
       "      <td>KABIGEN AB</td>\n",
       "      <td>1985</td>\n",
       "      <td>EPO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>579457</th>\n",
       "      <td>EP0155655A2</td>\n",
       "      <td>ASTELLAS PHARMA INC</td>\n",
       "      <td>1985</td>\n",
       "      <td>EPO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>579461</th>\n",
       "      <td>EP0158892A2</td>\n",
       "      <td>ASTELLAS PHARMA INC</td>\n",
       "      <td>1985</td>\n",
       "      <td>EPO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>540079</th>\n",
       "      <td>EP0297262A2</td>\n",
       "      <td>PROGEN BIOTECHNIK GMBH</td>\n",
       "      <td>1989</td>\n",
       "      <td>EPO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18273</th>\n",
       "      <td>WO9931260A2</td>\n",
       "      <td>MAX PLANCK SOCIETY</td>\n",
       "      <td>1999</td>\n",
       "      <td>WIPO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>504</th>\n",
       "      <td>WO9935271A2</td>\n",
       "      <td>CAMPOS GOMEZ JAVIER</td>\n",
       "      <td>1999</td>\n",
       "      <td>WIPO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41524</th>\n",
       "      <td>WO9943791A2</td>\n",
       "      <td>DIERL STEFAN (DE); RUETERJANS HEINZ (DE)</td>\n",
       "      <td>1999</td>\n",
       "      <td>WIPO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38922</th>\n",
       "      <td>WO9958639A2</td>\n",
       "      <td>TODD DANIEL (GB); WESTON JONATHAN (GB); AKZO N...</td>\n",
       "      <td>1999</td>\n",
       "      <td>WIPO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25410</th>\n",
       "      <td>WO9966052A1</td>\n",
       "      <td>BARBEYRON TRISTAN (FR); GOEMAR LAB SA (FR); FL...</td>\n",
       "      <td>1999</td>\n",
       "      <td>WIPO</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4779 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Application number                                          Applicant  \\\n",
       "534802       DE19859107A1                                  ROCHE DIAGNOSTICS   \n",
       "579459        EP0130166A1                                         KABIGEN AB   \n",
       "579457        EP0155655A2                                ASTELLAS PHARMA INC   \n",
       "579461        EP0158892A2                                ASTELLAS PHARMA INC   \n",
       "540079        EP0297262A2                             PROGEN BIOTECHNIK GMBH   \n",
       "...                   ...                                                ...   \n",
       "18273         WO9931260A2                                 MAX PLANCK SOCIETY   \n",
       "504           WO9935271A2                                CAMPOS GOMEZ JAVIER   \n",
       "41524         WO9943791A2           DIERL STEFAN (DE); RUETERJANS HEINZ (DE)   \n",
       "38922         WO9958639A2  TODD DANIEL (GB); WESTON JONATHAN (GB); AKZO N...   \n",
       "25410         WO9966052A1  BARBEYRON TRISTAN (FR); GOEMAR LAB SA (FR); FL...   \n",
       "\n",
       "        Year of application Patent system  \n",
       "534802                 2000           EPO  \n",
       "579459                 1985           EPO  \n",
       "579457                 1985           EPO  \n",
       "579461                 1985           EPO  \n",
       "540079                 1989           EPO  \n",
       "...                     ...           ...  \n",
       "18273                  1999          WIPO  \n",
       "504                    1999          WIPO  \n",
       "41524                  1999          WIPO  \n",
       "38922                  1999          WIPO  \n",
       "25410                  1999          WIPO  \n",
       "\n",
       "[4779 rows x 4 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mabpat_db_applications = mabpat_db_applications.sort_values(by='Application number', ascending=True).drop_duplicates(subset='Application number', keep='first')\n",
    "mabpat_db_applications"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### species table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species name</th>\n",
       "      <th>Domain</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Taxonomic source</th>\n",
       "      <th>Is Deep-sea</th>\n",
       "      <th>Deep-sea source</th>\n",
       "      <th>Is marine sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>44506</th>\n",
       "      <td>Acanthastrea echinata</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Cnidaria</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40917</th>\n",
       "      <td>Acanthopagrus schlegelii</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45925</th>\n",
       "      <td>Acanthus ebracteatus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Tracheophyta</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35907</th>\n",
       "      <td>Acaryochloris</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Wikipedia</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32178</th>\n",
       "      <td>Acaryochloris marina</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Cyanobacteria</td>\n",
       "      <td>Wikipedia</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39254</th>\n",
       "      <td>Zoarces viviparus</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Chordata</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25399</th>\n",
       "      <td>Zobellia galactanivorans</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>593469</th>\n",
       "      <td>Zobellia laminariae</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32713</th>\n",
       "      <td>Zostera marina</td>\n",
       "      <td>Eukaryota</td>\n",
       "      <td>Streptophyta</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>593249</th>\n",
       "      <td>Zunongwangia profunda</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1639 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    Species name     Domain         Phylum Taxonomic source  \\\n",
       "44506      Acanthastrea echinata  Eukaryota       Cnidaria            WoRMS   \n",
       "40917   Acanthopagrus schlegelii  Eukaryota       Chordata            WoRMS   \n",
       "45925       Acanthus ebracteatus  Eukaryota   Tracheophyta            WoRMS   \n",
       "35907              Acaryochloris   Bacteria  Cyanobacteria        Wikipedia   \n",
       "32178       Acaryochloris marina   Bacteria  Cyanobacteria        Wikipedia   \n",
       "...                          ...        ...            ...              ...   \n",
       "39254          Zoarces viviparus  Eukaryota       Chordata            WoRMS   \n",
       "25399   Zobellia galactanivorans   Bacteria  Bacteroidetes            WoRMS   \n",
       "593469       Zobellia laminariae   Bacteria  Bacteroidetes            WoRMS   \n",
       "32713             Zostera marina  Eukaryota   Streptophyta             NCBI   \n",
       "593249     Zunongwangia profunda   Bacteria  Bacteroidetes            WoRMS   \n",
       "\n",
       "        Is Deep-sea Deep-sea source  Is marine sequence  \n",
       "44506           0.0               0                   1  \n",
       "40917           0.0               0                   1  \n",
       "45925           0.0               0                   1  \n",
       "35907           0.0               0                   1  \n",
       "32178           0.0               0                   1  \n",
       "...             ...             ...                 ...  \n",
       "39254           0.0               0                   1  \n",
       "25399           0.0               0                   1  \n",
       "593469          0.0             NaN                   1  \n",
       "32713           0.0               0                   1  \n",
       "593249          0.0             NaN                   1  \n",
       "\n",
       "[1639 rows x 7 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#drop nans\n",
    "#mabpat_db_species = mabpat_db_species[mabpat_db_species['Is marine sequence']==1]\n",
    "#mabpat_db_species['Is marine species'] = mabpat_db_species['Is marine sequence']\n",
    "#del mabpat_db_species['Is marine sequence']\n",
    "mabpat_db_species = mabpat_db_species.sort_values(by='Species name', ascending=True).drop_duplicates(subset='Species name', keep='first')\n",
    "mabpat_db_species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# del mabpat_db_species['Is marine species']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Patents-sequences table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Application number</th>\n",
       "      <th>accession_num</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WO2004024931A2</td>\n",
       "      <td>CQ795562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>JP2010213712A</td>\n",
       "      <td>HV586656</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>JP2010213712A</td>\n",
       "      <td>HV586657</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>JP2010213712A</td>\n",
       "      <td>HV586658</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>JP2010213712A</td>\n",
       "      <td>HV586893</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>692348</th>\n",
       "      <td>EP2126131A2</td>\n",
       "      <td>HI462441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>692349</th>\n",
       "      <td>WO2010121818A1</td>\n",
       "      <td>HI646869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>692350</th>\n",
       "      <td>WO2010121818A1</td>\n",
       "      <td>HI646871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>692351</th>\n",
       "      <td>WO2010121818A1</td>\n",
       "      <td>HI646873</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>692352</th>\n",
       "      <td>WO2010121818A1</td>\n",
       "      <td>HI646875</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>557532 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Application number accession_num\n",
       "0          WO2004024931A2      CQ795562\n",
       "1           JP2010213712A      HV586656\n",
       "2           JP2010213712A      HV586657\n",
       "3           JP2010213712A      HV586658\n",
       "4           JP2010213712A      HV586893\n",
       "...                   ...           ...\n",
       "692348        EP2126131A2      HI462441\n",
       "692349     WO2010121818A1      HI646869\n",
       "692350     WO2010121818A1      HI646871\n",
       "692351     WO2010121818A1      HI646873\n",
       "692352     WO2010121818A1      HI646875\n",
       "\n",
       "[557532 rows x 2 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mabpat_db_app_sequences = mabpat_db_app_sequences.drop_duplicates()\n",
    "mabpat_db_app_sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Sequences table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession_num</th>\n",
       "      <th>Species name</th>\n",
       "      <th>GC content</th>\n",
       "      <th>Sequence length</th>\n",
       "      <th>Is marine sequence</th>\n",
       "      <th>Sequence status</th>\n",
       "      <th>Is protein-coding sequence</th>\n",
       "      <th>Is annotated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>540079</th>\n",
       "      <td>A01472</td>\n",
       "      <td>Physeter macrocephalus</td>\n",
       "      <td>51.826484</td>\n",
       "      <td>438</td>\n",
       "      <td>1</td>\n",
       "      <td>predicted</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>540078</th>\n",
       "      <td>A01473</td>\n",
       "      <td>Physeter macrocephalus</td>\n",
       "      <td>50.967742</td>\n",
       "      <td>465</td>\n",
       "      <td>1</td>\n",
       "      <td>predicted</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>551221</th>\n",
       "      <td>A01474</td>\n",
       "      <td>Trichechus manatus latirostris</td>\n",
       "      <td>55.483871</td>\n",
       "      <td>465</td>\n",
       "      <td>1</td>\n",
       "      <td>predicted</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>579457</th>\n",
       "      <td>A04570</td>\n",
       "      <td>Odobenus rosmarus divergens</td>\n",
       "      <td>53.809524</td>\n",
       "      <td>210</td>\n",
       "      <td>1</td>\n",
       "      <td>predicted</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>579458</th>\n",
       "      <td>A04571</td>\n",
       "      <td>Odobenus rosmarus divergens</td>\n",
       "      <td>53.809524</td>\n",
       "      <td>210</td>\n",
       "      <td>1</td>\n",
       "      <td>predicted</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22593</th>\n",
       "      <td>PA548965</td>\n",
       "      <td>Rhodothermus marinus</td>\n",
       "      <td>62.593783</td>\n",
       "      <td>1866</td>\n",
       "      <td>1</td>\n",
       "      <td>observed</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22594</th>\n",
       "      <td>PA548966</td>\n",
       "      <td>Rhodothermus marinus</td>\n",
       "      <td>62.593783</td>\n",
       "      <td>1866</td>\n",
       "      <td>1</td>\n",
       "      <td>observed</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10851</th>\n",
       "      <td>PA548992</td>\n",
       "      <td>Pyrococcus furiosus</td>\n",
       "      <td>47.637457</td>\n",
       "      <td>2328</td>\n",
       "      <td>1</td>\n",
       "      <td>observed</td>\n",
       "      <td>True</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38938</th>\n",
       "      <td>PA550739</td>\n",
       "      <td>Anabaena cylindrica pcc 7122</td>\n",
       "      <td>45.147059</td>\n",
       "      <td>1360</td>\n",
       "      <td>1</td>\n",
       "      <td>observed</td>\n",
       "      <td>True</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38939</th>\n",
       "      <td>PA550742</td>\n",
       "      <td>Anabaena cylindrica pcc 7122</td>\n",
       "      <td>38.271605</td>\n",
       "      <td>1053</td>\n",
       "      <td>1</td>\n",
       "      <td>observed</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>104467 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       accession_num                    Species name  GC content  \\\n",
       "540079        A01472          Physeter macrocephalus   51.826484   \n",
       "540078        A01473          Physeter macrocephalus   50.967742   \n",
       "551221        A01474  Trichechus manatus latirostris   55.483871   \n",
       "579457        A04570     Odobenus rosmarus divergens   53.809524   \n",
       "579458        A04571     Odobenus rosmarus divergens   53.809524   \n",
       "...              ...                             ...         ...   \n",
       "22593       PA548965            Rhodothermus marinus   62.593783   \n",
       "22594       PA548966            Rhodothermus marinus   62.593783   \n",
       "10851       PA548992             Pyrococcus furiosus   47.637457   \n",
       "38938       PA550739    Anabaena cylindrica pcc 7122   45.147059   \n",
       "38939       PA550742    Anabaena cylindrica pcc 7122   38.271605   \n",
       "\n",
       "        Sequence length  Is marine sequence Sequence status  \\\n",
       "540079              438                   1       predicted   \n",
       "540078              465                   1       predicted   \n",
       "551221              465                   1       predicted   \n",
       "579457              210                   1       predicted   \n",
       "579458              210                   1       predicted   \n",
       "...                 ...                 ...             ...   \n",
       "22593              1866                   1        observed   \n",
       "22594              1866                   1        observed   \n",
       "10851              2328                   1        observed   \n",
       "38938              1360                   1        observed   \n",
       "38939              1053                   1        observed   \n",
       "\n",
       "        Is protein-coding sequence  Is annotated  \n",
       "540079                        True             1  \n",
       "540078                        True             1  \n",
       "551221                        True             1  \n",
       "579457                        True             1  \n",
       "579458                        True             1  \n",
       "...                            ...           ...  \n",
       "22593                         True             0  \n",
       "22594                        False             0  \n",
       "10851                         True             1  \n",
       "38938                         True             0  \n",
       "38939                        False             0  \n",
       "\n",
       "[104467 rows x 8 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mabpat_db_sequences = mabpat_db_sequences.sort_values(by='accession_num', ascending=True).drop_duplicates(subset='accession_num', keep='first')\n",
    "mabpat_db_sequences"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Annotated marine proteins table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/user/1024/ipykernel_3420031/169507504.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mabpat_db_marine_proteins_annotated_functions_no_nans['annotation_source'] = mabpat_db_marine_proteins_annotated_functions_no_nans.sseqid.apply(lambda x: x[0:2].strip())\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession_num</th>\n",
       "      <th>f_header</th>\n",
       "      <th>sseqid</th>\n",
       "      <th>stitle</th>\n",
       "      <th>pident</th>\n",
       "      <th>evalue</th>\n",
       "      <th>qcovs</th>\n",
       "      <th>annotation_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>12703</th>\n",
       "      <td>FB793687</td>\n",
       "      <td>36530217_pyrobaculum</td>\n",
       "      <td>sp|Q8ZYE2|RS2_PYRAE</td>\n",
       "      <td>sp|Q8ZYE2|RS2_PYRAE Small ribosomal subunit pr...</td>\n",
       "      <td>100.0</td>\n",
       "      <td>6.870000e-153</td>\n",
       "      <td>99.5</td>\n",
       "      <td>sp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534412</th>\n",
       "      <td>KH326754</td>\n",
       "      <td>11132701_unknown.</td>\n",
       "      <td>sp|Q9KR66|SIAM_VIBCH</td>\n",
       "      <td>sp|Q9KR66|SIAM_VIBCH Sialic acid TRAP transpor...</td>\n",
       "      <td>95.1</td>\n",
       "      <td>4.520000e-277</td>\n",
       "      <td>99.8</td>\n",
       "      <td>sp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534413</th>\n",
       "      <td>GZ076609</td>\n",
       "      <td>48485655_unknown.</td>\n",
       "      <td>sp|P01556|CHTB_VIBCH</td>\n",
       "      <td>sp|P01556|CHTB_VIBCH Cholera enterotoxin subun...</td>\n",
       "      <td>98.1</td>\n",
       "      <td>5.020000e-69</td>\n",
       "      <td>95.4</td>\n",
       "      <td>sp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534414</th>\n",
       "      <td>GZ076610</td>\n",
       "      <td>48485656_unknown.</td>\n",
       "      <td>sp|P01556|CHTB_VIBCH</td>\n",
       "      <td>sp|P01556|CHTB_VIBCH Cholera enterotoxin subun...</td>\n",
       "      <td>96.1</td>\n",
       "      <td>6.860000e-67</td>\n",
       "      <td>95.4</td>\n",
       "      <td>sp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>534415</th>\n",
       "      <td>HJ399034</td>\n",
       "      <td>54764162_unknown.</td>\n",
       "      <td>sp|A5F5G0|RPIA_VIBC3</td>\n",
       "      <td>sp|A5F5G0|RPIA_VIBC3 Ribose-5-phosphate isomer...</td>\n",
       "      <td>96.8</td>\n",
       "      <td>1.420000e-150</td>\n",
       "      <td>99.5</td>\n",
       "      <td>sp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>548211</th>\n",
       "      <td>MY231046</td>\n",
       "      <td>30208840_unknown.</td>\n",
       "      <td>tr|A0A2Y9E2H6|A0A2Y9E2H6_TRIMA</td>\n",
       "      <td>tr|A0A2Y9E2H6|A0A2Y9E2H6_TRIMA Rho GTPase-acti...</td>\n",
       "      <td>98.4</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>99.9</td>\n",
       "      <td>tr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>548210</th>\n",
       "      <td>MY231110</td>\n",
       "      <td>30208904_unknown.</td>\n",
       "      <td>tr|A0A2Y9E5G9|A0A2Y9E5G9_TRIMA</td>\n",
       "      <td>tr|A0A2Y9E5G9|A0A2Y9E5G9_TRIMA Aryl hydrocarbo...</td>\n",
       "      <td>98.2</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>98.3</td>\n",
       "      <td>tr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>548209</th>\n",
       "      <td>MY231091</td>\n",
       "      <td>30208885_unknown.</td>\n",
       "      <td>tr|A0A2Y9QLM4|A0A2Y9QLM4_TRIMA</td>\n",
       "      <td>tr|A0A2Y9QLM4|A0A2Y9QLM4_TRIMA ADP-ribosylatio...</td>\n",
       "      <td>100.0</td>\n",
       "      <td>3.460000e-140</td>\n",
       "      <td>99.5</td>\n",
       "      <td>tr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>548215</th>\n",
       "      <td>MY231166</td>\n",
       "      <td>30208960_unknown.</td>\n",
       "      <td>tr|A0A2Y9E1L6|A0A2Y9E1L6_TRIMA</td>\n",
       "      <td>tr|A0A2Y9E1L6|A0A2Y9E1L6_TRIMA Acid-sensing io...</td>\n",
       "      <td>99.6</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>99.8</td>\n",
       "      <td>tr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>594922</th>\n",
       "      <td>HL838923</td>\n",
       "      <td>2596833_unknown.</td>\n",
       "      <td>tr|G0AV56|G0AV56_9GAMM</td>\n",
       "      <td>tr|G0AV56|G0AV56_9GAMM Anti-ECFsigma factor, C...</td>\n",
       "      <td>96.7</td>\n",
       "      <td>1.140000e-167</td>\n",
       "      <td>99.6</td>\n",
       "      <td>tr</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>71149 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       accession_num              f_header                          sseqid  \\\n",
       "12703       FB793687  36530217_pyrobaculum             sp|Q8ZYE2|RS2_PYRAE   \n",
       "534412      KH326754     11132701_unknown.            sp|Q9KR66|SIAM_VIBCH   \n",
       "534413      GZ076609     48485655_unknown.            sp|P01556|CHTB_VIBCH   \n",
       "534414      GZ076610     48485656_unknown.            sp|P01556|CHTB_VIBCH   \n",
       "534415      HJ399034     54764162_unknown.            sp|A5F5G0|RPIA_VIBC3   \n",
       "...              ...                   ...                             ...   \n",
       "548211      MY231046     30208840_unknown.  tr|A0A2Y9E2H6|A0A2Y9E2H6_TRIMA   \n",
       "548210      MY231110     30208904_unknown.  tr|A0A2Y9E5G9|A0A2Y9E5G9_TRIMA   \n",
       "548209      MY231091     30208885_unknown.  tr|A0A2Y9QLM4|A0A2Y9QLM4_TRIMA   \n",
       "548215      MY231166     30208960_unknown.  tr|A0A2Y9E1L6|A0A2Y9E1L6_TRIMA   \n",
       "594922      HL838923      2596833_unknown.          tr|G0AV56|G0AV56_9GAMM   \n",
       "\n",
       "                                                   stitle  pident  \\\n",
       "12703   sp|Q8ZYE2|RS2_PYRAE Small ribosomal subunit pr...   100.0   \n",
       "534412  sp|Q9KR66|SIAM_VIBCH Sialic acid TRAP transpor...    95.1   \n",
       "534413  sp|P01556|CHTB_VIBCH Cholera enterotoxin subun...    98.1   \n",
       "534414  sp|P01556|CHTB_VIBCH Cholera enterotoxin subun...    96.1   \n",
       "534415  sp|A5F5G0|RPIA_VIBC3 Ribose-5-phosphate isomer...    96.8   \n",
       "...                                                   ...     ...   \n",
       "548211  tr|A0A2Y9E2H6|A0A2Y9E2H6_TRIMA Rho GTPase-acti...    98.4   \n",
       "548210  tr|A0A2Y9E5G9|A0A2Y9E5G9_TRIMA Aryl hydrocarbo...    98.2   \n",
       "548209  tr|A0A2Y9QLM4|A0A2Y9QLM4_TRIMA ADP-ribosylatio...   100.0   \n",
       "548215  tr|A0A2Y9E1L6|A0A2Y9E1L6_TRIMA Acid-sensing io...    99.6   \n",
       "594922  tr|G0AV56|G0AV56_9GAMM Anti-ECFsigma factor, C...    96.7   \n",
       "\n",
       "               evalue  qcovs annotation_source  \n",
       "12703   6.870000e-153   99.5                sp  \n",
       "534412  4.520000e-277   99.8                sp  \n",
       "534413   5.020000e-69   95.4                sp  \n",
       "534414   6.860000e-67   95.4                sp  \n",
       "534415  1.420000e-150   99.5                sp  \n",
       "...               ...    ...               ...  \n",
       "548211   0.000000e+00   99.9                tr  \n",
       "548210   0.000000e+00   98.3                tr  \n",
       "548209  3.460000e-140   99.5                tr  \n",
       "548215   0.000000e+00   99.8                tr  \n",
       "594922  1.140000e-167   99.6                tr  \n",
       "\n",
       "[71149 rows x 8 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Drop nans\n",
    "mabpat_db_marine_proteins_annotated_functions_no_nans = mabpat_db_marine_proteins_annotated_functions[mabpat_db_marine_proteins_annotated_functions['f_header'].notna()]\n",
    "mabpat_db_marine_proteins_annotated_functions_no_nans['annotation_source'] = mabpat_db_marine_proteins_annotated_functions_no_nans.sseqid.apply(lambda x: x[0:2].strip())\n",
    "#remove 2151 duplicates\n",
    "mabpat_db_marine_proteins_annotated_functions_no_nans = mabpat_db_marine_proteins_annotated_functions_no_nans.sort_values(by='annotation_source', ascending=True).drop_duplicates(subset='accession_num', keep='first')\n",
    "mabpat_db_marine_proteins_annotated_functions_no_nans"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Non-marine sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession_num</th>\n",
       "      <th>Species name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>45985</th>\n",
       "      <td>HV037167</td>\n",
       "      <td>Bacillus licheniformis dsm 13 = atcc 14580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45986</th>\n",
       "      <td>HV037168</td>\n",
       "      <td>Bacillus pumilus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45987</th>\n",
       "      <td>HV037169</td>\n",
       "      <td>Acetivibrio thermocellus atcc 27405</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45988</th>\n",
       "      <td>HV037172</td>\n",
       "      <td>Thermoanaerobacterium sp.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45989</th>\n",
       "      <td>HV037173</td>\n",
       "      <td>Bacillus sp. nrrl b-14911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>692348</th>\n",
       "      <td>HI462441</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>692349</th>\n",
       "      <td>HI646869</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>692350</th>\n",
       "      <td>HI646871</td>\n",
       "      <td>Arabidopsis thaliana</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>692351</th>\n",
       "      <td>HI646873</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>692352</th>\n",
       "      <td>HI646875</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>453114 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       accession_num                                Species name\n",
       "45985       HV037167  Bacillus licheniformis dsm 13 = atcc 14580\n",
       "45986       HV037168                            Bacillus pumilus\n",
       "45987       HV037169         Acetivibrio thermocellus atcc 27405\n",
       "45988       HV037172                   Thermoanaerobacterium sp.\n",
       "45989       HV037173                   Bacillus sp. nrrl b-14911\n",
       "...              ...                                         ...\n",
       "692348      HI462441                                Homo sapiens\n",
       "692349      HI646869                        Arabidopsis thaliana\n",
       "692350      HI646871                        Arabidopsis thaliana\n",
       "692351      HI646873                                Homo sapiens\n",
       "692352      HI646875                                Homo sapiens\n",
       "\n",
       "[453114 rows x 2 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mabpat_db_cleaned_non_marine"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save csv tables AND Create SQL database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "mabpat_db_applicants.to_csv('/home/erikz/data/mabpat/dbs/final_tables/01-applicants-table.csv', index=False)\n",
    "mabpat_db_applications.to_csv('/home/erikz/data/mabpat/dbs/final_tables/02-applications-table.csv', index=False)\n",
    "mabpat_db_species.to_csv('/home/erikz/data/mabpat/dbs/final_tables/03-marine-species-table.csv', index=False)\n",
    "mabpat_db_app_sequences.to_csv('/home/erikz/data/mabpat/dbs/final_tables/04-patents-sequences-table.csv', index=False)\n",
    "mabpat_db_sequences.to_csv('/home/erikz/data/mabpat/dbs/final_tables/05-sequences-table.csv', index=False)\n",
    "mabpat_db_marine_proteins_annotated_functions_no_nans.to_csv('/home/erikz/data/mabpat/dbs/final_tables/06-annotated-marine-proteins-table.csv', index=False)\n",
    "mabpat_db_cleaned_non_marine.to_csv('/home/erikz/data/mabpat/dbs/final_tables/07-non-marine-sequences-table.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ALL Tables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [],
   "source": [
    "mabpat_db_merged1 = pd.merge(mabpat_db_applicants,mabpat_db_applications,how='outer',on='Applicant').drop_duplicates()\n",
    "\n",
    "mabpat_db_merged2 = pd.merge(mabpat_db_sequences,mabpat_db_species,how='outer',on='Species name').drop_duplicates()\n",
    "\n",
    "mabpat_db_merged2 = pd.merge(mabpat_db_merged2,mabpat_db_marine_proteins_annotated_functions_no_nans,how='outer',on='accession_num').drop_duplicates()\n",
    "\n",
    "mabpat_db_merged2 = pd.merge(mabpat_db_merged2,mabpat_db_app_sequences,how='outer',on='accession_num').drop_duplicates()\n",
    "\n",
    "mabpat_db_merged = pd.merge(mabpat_db_merged1,mabpat_db_merged2,how='outer',on='Application number').drop_duplicates()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Applicant</th>\n",
       "      <th>Applicant type</th>\n",
       "      <th>Applicant country</th>\n",
       "      <th>Application number</th>\n",
       "      <th>Year of application</th>\n",
       "      <th>Patent system</th>\n",
       "      <th>accession_num</th>\n",
       "      <th>Species name</th>\n",
       "      <th>GC content</th>\n",
       "      <th>Sequence length</th>\n",
       "      <th>...</th>\n",
       "      <th>Taxonomic source</th>\n",
       "      <th>Is Deep-sea</th>\n",
       "      <th>Deep-sea source</th>\n",
       "      <th>f_header</th>\n",
       "      <th>sseqid</th>\n",
       "      <th>stitle</th>\n",
       "      <th>pident</th>\n",
       "      <th>evalue</th>\n",
       "      <th>qcovs</th>\n",
       "      <th>annotation_source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1GLOBE BIOMEDICAL CO., LTD.|1GLOBE HEALTH INST...</td>\n",
       "      <td>NATIONAL</td>\n",
       "      <td>HONG KONG SAR</td>\n",
       "      <td>KR1020210046725A</td>\n",
       "      <td>2021.0</td>\n",
       "      <td>KIPO</td>\n",
       "      <td>OF021152</td>\n",
       "      <td>Pseudoalteromonas caenipelagi</td>\n",
       "      <td>49.844237</td>\n",
       "      <td>321.0</td>\n",
       "      <td>...</td>\n",
       "      <td>Wikipedia</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>30791729_synthetic</td>\n",
       "      <td>tr|A0A849VIE7|A0A849VIE7_9GAMM</td>\n",
       "      <td>tr|A0A849VIE7|A0A849VIE7_9GAMM Ig-like domain-...</td>\n",
       "      <td>96.2</td>\n",
       "      <td>3.510000e-62</td>\n",
       "      <td>99.1</td>\n",
       "      <td>tr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3-DIMENSIONAL PHARMACEUTICALS, INC.; EXTON, PA</td>\n",
       "      <td>NATIONAL</td>\n",
       "      <td>US</td>\n",
       "      <td>US6737404A</td>\n",
       "      <td>2004.0</td>\n",
       "      <td>USPTO</td>\n",
       "      <td>AR540675</td>\n",
       "      <td>Physeter macrocephalus</td>\n",
       "      <td>55.486542</td>\n",
       "      <td>483.0</td>\n",
       "      <td>...</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>WoRDSS</td>\n",
       "      <td>21699286_unknown.</td>\n",
       "      <td>tr|A0A2Y9F7B8|A0A2Y9F7B8_PHYMC</td>\n",
       "      <td>tr|A0A2Y9F7B8|A0A2Y9F7B8_PHYMC Fibroblast grow...</td>\n",
       "      <td>98.7</td>\n",
       "      <td>6.210000e-111</td>\n",
       "      <td>96.3</td>\n",
       "      <td>tr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A GREY CRAIG,DAVID GRIFFEN,BALDOMERO M OLIVERA...</td>\n",
       "      <td>NONE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>JP2011024579A</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>JPO</td>\n",
       "      <td>HV600304</td>\n",
       "      <td>Conus geographus</td>\n",
       "      <td>17.647059</td>\n",
       "      <td>17.0</td>\n",
       "      <td>...</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>A GREY CRAIG,DAVID GRIFFEN,BALDOMERO M OLIVERA...</td>\n",
       "      <td>NONE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>JP2011024579A</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>JPO</td>\n",
       "      <td>HV600305</td>\n",
       "      <td>Conus geographus</td>\n",
       "      <td>51.515152</td>\n",
       "      <td>231.0</td>\n",
       "      <td>...</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "      <td>3592600_conus</td>\n",
       "      <td>sp|Q9XYR5|CONG_CONGE</td>\n",
       "      <td>sp|Q9XYR5|CONG_CONGE Contulakin-G OS=Conus geo...</td>\n",
       "      <td>98.7</td>\n",
       "      <td>1.230000e-49</td>\n",
       "      <td>98.7</td>\n",
       "      <td>sp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AARHUS UNIVERSITET (DK)</td>\n",
       "      <td>UNIVERSITY</td>\n",
       "      <td>DENMARK</td>\n",
       "      <td>WO2010094288A1</td>\n",
       "      <td>2010.0</td>\n",
       "      <td>WIPO</td>\n",
       "      <td>HH263082</td>\n",
       "      <td>Methanosarcina mazei</td>\n",
       "      <td>56.318681</td>\n",
       "      <td>1092.0</td>\n",
       "      <td>...</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>53160624_synthetic</td>\n",
       "      <td>tr|A0A0F8J5K8|A0A0F8J5K8_METMZ</td>\n",
       "      <td>tr|A0A0F8J5K8|A0A0F8J5K8_METMZ Lac repressor O...</td>\n",
       "      <td>100.0</td>\n",
       "      <td>4.760000e-246</td>\n",
       "      <td>98.6</td>\n",
       "      <td>tr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104464</th>\n",
       "      <td>ZALTSMAN, ADI</td>\n",
       "      <td>NONE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>KR1020150093721A</td>\n",
       "      <td>2015.0</td>\n",
       "      <td>KIPO</td>\n",
       "      <td>LG087077</td>\n",
       "      <td>Richelia sinica fachb-800</td>\n",
       "      <td>41.163311</td>\n",
       "      <td>894.0</td>\n",
       "      <td>...</td>\n",
       "      <td>NCBI</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12951498_synthetic</td>\n",
       "      <td>tr|A0A975T4R8|A0A975T4R8_9NOST</td>\n",
       "      <td>tr|A0A975T4R8|A0A975T4R8_9NOST Nitrogenase iro...</td>\n",
       "      <td>98.3</td>\n",
       "      <td>2.620000e-206</td>\n",
       "      <td>99.7</td>\n",
       "      <td>tr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104465</th>\n",
       "      <td>ZENSUN (SHANGHAI) SCI &amp; TECH CO., LTD</td>\n",
       "      <td>NATIONAL</td>\n",
       "      <td>CHINA</td>\n",
       "      <td>US8609620B2</td>\n",
       "      <td>2013.0</td>\n",
       "      <td>USPTO</td>\n",
       "      <td>HK343917</td>\n",
       "      <td>Odobenus rosmarus divergens</td>\n",
       "      <td>46.540881</td>\n",
       "      <td>159.0</td>\n",
       "      <td>...</td>\n",
       "      <td>Wikipedia</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1101828_unknown.</td>\n",
       "      <td>tr|A0A2U3WRP2|A0A2U3WRP2_ODORO</td>\n",
       "      <td>tr|A0A2U3WRP2|A0A2U3WRP2_ODORO Pro-neuregulin-...</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1.650000e-33</td>\n",
       "      <td>100.0</td>\n",
       "      <td>tr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104466</th>\n",
       "      <td>ZENSUN (SHANGHAI) SCI &amp; TECH CO., LTD</td>\n",
       "      <td>NATIONAL</td>\n",
       "      <td>CHINA</td>\n",
       "      <td>US8609620B2</td>\n",
       "      <td>2013.0</td>\n",
       "      <td>USPTO</td>\n",
       "      <td>HK343918</td>\n",
       "      <td>Odobenus rosmarus divergens</td>\n",
       "      <td>46.060606</td>\n",
       "      <td>165.0</td>\n",
       "      <td>...</td>\n",
       "      <td>Wikipedia</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1101829_unknown.</td>\n",
       "      <td>tr|A0A2U3WRP2|A0A2U3WRP2_ODORO</td>\n",
       "      <td>tr|A0A2U3WRP2|A0A2U3WRP2_ODORO Pro-neuregulin-...</td>\n",
       "      <td>100.0</td>\n",
       "      <td>1.580000e-34</td>\n",
       "      <td>98.2</td>\n",
       "      <td>tr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104467</th>\n",
       "      <td>ZUCHEM INC [US]</td>\n",
       "      <td>NATIONAL</td>\n",
       "      <td>US</td>\n",
       "      <td>EP2606141A2</td>\n",
       "      <td>2013.0</td>\n",
       "      <td>EPO</td>\n",
       "      <td>JC366125</td>\n",
       "      <td>Pyrococcus furiosus</td>\n",
       "      <td>38.810198</td>\n",
       "      <td>1059.0</td>\n",
       "      <td>...</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2018</td>\n",
       "      <td>8055890_pyrococcus</td>\n",
       "      <td>sp|Q9HHB6|GAL1_PYRFU</td>\n",
       "      <td>sp|Q9HHB6|GAL1_PYRFU Galactokinase OS=Pyrococc...</td>\n",
       "      <td>100.0</td>\n",
       "      <td>4.370000e-253</td>\n",
       "      <td>99.7</td>\n",
       "      <td>sp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104468</th>\n",
       "      <td>ZUCHEM INC [US]</td>\n",
       "      <td>NATIONAL</td>\n",
       "      <td>US</td>\n",
       "      <td>EP2606141A2</td>\n",
       "      <td>2013.0</td>\n",
       "      <td>EPO</td>\n",
       "      <td>JC366128</td>\n",
       "      <td>Pyrococcus furiosus</td>\n",
       "      <td>39.429929</td>\n",
       "      <td>1263.0</td>\n",
       "      <td>...</td>\n",
       "      <td>WoRMS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2018</td>\n",
       "      <td>8055893_pyrococcus</td>\n",
       "      <td>tr|A0A5C0XMC6|A0A5C0XMC6_PYRFU</td>\n",
       "      <td>tr|A0A5C0XMC6|A0A5C0XMC6_PYRFU Bifunctional pr...</td>\n",
       "      <td>100.0</td>\n",
       "      <td>2.240000e-269</td>\n",
       "      <td>99.8</td>\n",
       "      <td>tr</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>104469 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                Applicant Applicant type  \\\n",
       "0       1GLOBE BIOMEDICAL CO., LTD.|1GLOBE HEALTH INST...       NATIONAL   \n",
       "1          3-DIMENSIONAL PHARMACEUTICALS, INC.; EXTON, PA       NATIONAL   \n",
       "2       A GREY CRAIG,DAVID GRIFFEN,BALDOMERO M OLIVERA...           NONE   \n",
       "3       A GREY CRAIG,DAVID GRIFFEN,BALDOMERO M OLIVERA...           NONE   \n",
       "4                                 AARHUS UNIVERSITET (DK)     UNIVERSITY   \n",
       "...                                                   ...            ...   \n",
       "104464                                      ZALTSMAN, ADI           NONE   \n",
       "104465              ZENSUN (SHANGHAI) SCI & TECH CO., LTD       NATIONAL   \n",
       "104466              ZENSUN (SHANGHAI) SCI & TECH CO., LTD       NATIONAL   \n",
       "104467                                    ZUCHEM INC [US]       NATIONAL   \n",
       "104468                                    ZUCHEM INC [US]       NATIONAL   \n",
       "\n",
       "       Applicant country Application number  Year of application  \\\n",
       "0          HONG KONG SAR   KR1020210046725A               2021.0   \n",
       "1                     US         US6737404A               2004.0   \n",
       "2                    NaN      JP2011024579A               2011.0   \n",
       "3                    NaN      JP2011024579A               2011.0   \n",
       "4                DENMARK     WO2010094288A1               2010.0   \n",
       "...                  ...                ...                  ...   \n",
       "104464               NaN   KR1020150093721A               2015.0   \n",
       "104465             CHINA        US8609620B2               2013.0   \n",
       "104466             CHINA        US8609620B2               2013.0   \n",
       "104467                US        EP2606141A2               2013.0   \n",
       "104468                US        EP2606141A2               2013.0   \n",
       "\n",
       "       Patent system accession_num                   Species name  GC content  \\\n",
       "0               KIPO      OF021152  Pseudoalteromonas caenipelagi   49.844237   \n",
       "1              USPTO      AR540675         Physeter macrocephalus   55.486542   \n",
       "2                JPO      HV600304               Conus geographus   17.647059   \n",
       "3                JPO      HV600305               Conus geographus   51.515152   \n",
       "4               WIPO      HH263082           Methanosarcina mazei   56.318681   \n",
       "...              ...           ...                            ...         ...   \n",
       "104464          KIPO      LG087077      Richelia sinica fachb-800   41.163311   \n",
       "104465         USPTO      HK343917    Odobenus rosmarus divergens   46.540881   \n",
       "104466         USPTO      HK343918    Odobenus rosmarus divergens   46.060606   \n",
       "104467           EPO      JC366125            Pyrococcus furiosus   38.810198   \n",
       "104468           EPO      JC366128            Pyrococcus furiosus   39.429929   \n",
       "\n",
       "        Sequence length  ...  Taxonomic source Is Deep-sea Deep-sea source  \\\n",
       "0                 321.0  ...         Wikipedia         0.0             NaN   \n",
       "1                 483.0  ...             WoRMS         1.0          WoRDSS   \n",
       "2                  17.0  ...             WoRMS         0.0               0   \n",
       "3                 231.0  ...             WoRMS         0.0               0   \n",
       "4                1092.0  ...             WoRMS         0.0             NaN   \n",
       "...                 ...  ...               ...         ...             ...   \n",
       "104464            894.0  ...              NCBI         0.0             NaN   \n",
       "104465            159.0  ...         Wikipedia         0.0             NaN   \n",
       "104466            165.0  ...         Wikipedia         0.0             NaN   \n",
       "104467           1059.0  ...             WoRMS         1.0            2018   \n",
       "104468           1263.0  ...             WoRMS         1.0            2018   \n",
       "\n",
       "                  f_header                          sseqid  \\\n",
       "0       30791729_synthetic  tr|A0A849VIE7|A0A849VIE7_9GAMM   \n",
       "1        21699286_unknown.  tr|A0A2Y9F7B8|A0A2Y9F7B8_PHYMC   \n",
       "2                      NaN                             NaN   \n",
       "3            3592600_conus            sp|Q9XYR5|CONG_CONGE   \n",
       "4       53160624_synthetic  tr|A0A0F8J5K8|A0A0F8J5K8_METMZ   \n",
       "...                    ...                             ...   \n",
       "104464  12951498_synthetic  tr|A0A975T4R8|A0A975T4R8_9NOST   \n",
       "104465    1101828_unknown.  tr|A0A2U3WRP2|A0A2U3WRP2_ODORO   \n",
       "104466    1101829_unknown.  tr|A0A2U3WRP2|A0A2U3WRP2_ODORO   \n",
       "104467  8055890_pyrococcus            sp|Q9HHB6|GAL1_PYRFU   \n",
       "104468  8055893_pyrococcus  tr|A0A5C0XMC6|A0A5C0XMC6_PYRFU   \n",
       "\n",
       "                                                   stitle pident  \\\n",
       "0       tr|A0A849VIE7|A0A849VIE7_9GAMM Ig-like domain-...   96.2   \n",
       "1       tr|A0A2Y9F7B8|A0A2Y9F7B8_PHYMC Fibroblast grow...   98.7   \n",
       "2                                                     NaN    NaN   \n",
       "3       sp|Q9XYR5|CONG_CONGE Contulakin-G OS=Conus geo...   98.7   \n",
       "4       tr|A0A0F8J5K8|A0A0F8J5K8_METMZ Lac repressor O...  100.0   \n",
       "...                                                   ...    ...   \n",
       "104464  tr|A0A975T4R8|A0A975T4R8_9NOST Nitrogenase iro...   98.3   \n",
       "104465  tr|A0A2U3WRP2|A0A2U3WRP2_ODORO Pro-neuregulin-...  100.0   \n",
       "104466  tr|A0A2U3WRP2|A0A2U3WRP2_ODORO Pro-neuregulin-...  100.0   \n",
       "104467  sp|Q9HHB6|GAL1_PYRFU Galactokinase OS=Pyrococc...  100.0   \n",
       "104468  tr|A0A5C0XMC6|A0A5C0XMC6_PYRFU Bifunctional pr...  100.0   \n",
       "\n",
       "               evalue  qcovs annotation_source  \n",
       "0        3.510000e-62   99.1                tr  \n",
       "1       6.210000e-111   96.3                tr  \n",
       "2                 NaN    NaN               NaN  \n",
       "3        1.230000e-49   98.7                sp  \n",
       "4       4.760000e-246   98.6                tr  \n",
       "...               ...    ...               ...  \n",
       "104464  2.620000e-206   99.7                tr  \n",
       "104465   1.650000e-33  100.0                tr  \n",
       "104466   1.580000e-34   98.2                tr  \n",
       "104467  4.370000e-253   99.7                sp  \n",
       "104468  2.240000e-269   99.8                tr  \n",
       "\n",
       "[104469 rows x 26 columns]"
      ]
     },
     "execution_count": 213,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mabpat_db_merged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [],
   "source": [
    "mabpat_db_merged.to_csv('/home/erikz/data/mabpat/dbs/MABPAT_050224_COMPLETED_MARINE_ONLY.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SQL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [],
   "source": [
    "con = sqlite3.connect(\"/home/erikz/data/mabpat/dbs/final_sql/mabpat_reduced.sqlite\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "71149"
      ]
     },
     "execution_count": 184,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#mabpat_db_cleaned.to_sql('one single table', con)\n",
    "mabpat_db_applicants.to_sql('Applicants', con)\n",
    "mabpat_db_applications.to_sql('Applications', con)\n",
    "mabpat_db_species.to_sql('MarineSpecies', con)\n",
    "mabpat_db_app_sequences.to_sql('Applications_Sequences', con)\n",
    "mabpat_db_sequences.to_sql('Sequences', con)\n",
    "mabpat_db_marine_proteins_annotated_functions_no_nans.to_sql('AnnotatedMarineProteins', con)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {},
   "outputs": [],
   "source": [
    "# We close the connection\n",
    "con.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Assign primary keys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Connect to the SQLite database\n",
    "conn = sqlite3.connect('/home/erikz/data/mabpat/dbs/final_sql/mabpat_reduced.sqlite')\n",
    "cursor = conn.cursor()\n",
    "\n",
    "# Step 1: Create a new table with a primary key\n",
    "'Applicant', 'Applicant_type', 'Applicant country'\n",
    "cursor.execute('''\n",
    "CREATE TABLE Patent_Applicants (\n",
    "    Applicant varchar(255) NOT NULL,\n",
    "    Applicant_type varchar(255),\n",
    "    Address varchar(255),\n",
    "    City varchar(255),\n",
    "    PRIMARY KEY (EmployeeID)\n",
    ");\n",
    "''')\n",
    "\n",
    "# Step 2: Copy data from the existing table to the new table\n",
    "cursor.execute('''\n",
    "INSERT INTO new_Employees (EmployeeID, LastName, FirstName, Address, City)\n",
    "SELECT EmployeeID, LastName, FirstName, Address, City FROM Employees;\n",
    "''')\n",
    "\n",
    "# Step 3: Drop the old table\n",
    "cursor.execute('DROP TABLE Employees;')\n",
    "\n",
    "# Step 4: Rename the new table to the original table name\n",
    "cursor.execute('ALTER TABLE new_Employees RENAME TO Employees;')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "mgr",
   "language": "python",
   "name": "mgr"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
